### PR TITLE
add MobileNetV3

### PR DESCRIPTION
**Pull request by SigureMo**
### PR types

New features

### PR changes

APIs

### Describe

- 向 paddle.vision.models 添加以下模型

  - MobileNetV3Small
  - MobileNetV3Large
  - mobilenet_v3_small
  - mobilenet_v3_large

### Tasks

- [x] 支持预训练模型
- [x] 单元测试
- [x] 中英文文档
   - [x] 中文文档：https://github.com/PaddlePaddle/docs/pull/4176
   - [x] 英文文档（代码 DocString）
- [x] 精度自测，如下：

### Performance <sup>updated</sup>

AI Studio 测试详情：<https://aistudio.baidu.com/studio/project/partial/verify/3294252/037497a8fe694d89954b1c6e9bf274dd>
基准参考：<https://github.com/PaddlePaddle/PaddleClas/blob/release/2.2/docs/en/ImageNet_models_en.md>

|Model|Top-1|Top-5|
|-|-|-|
|mobilenet_v3_small_1_0|0.67668<span style="color:blue;"><sub>(=)</sub></span>|0.87402<span style="color:blue;"><sub>(=)</sub></span>|
|mobilenet_v3_large_1_0|0.74042<span style="color:blue;"><sub>(=)</sub></span>|0.91340<span style="color:blue;"><sub>(=)</sub></span>|

> 括号中为以 torchvision 性能基准为参考的偏差值

虽然有很多不同 scale 的预训练模型，但这里参考了 paddle 的 mobilenet_v1 和 mobilenet_v2 使用参数 scale 调节 scale，并没有为每个 scale 暴露一个 API，同时也与 torchvision 暴露的 mobilenet_v3_small 和 mobilenet_v3_large 对齐～不过我不太清楚这样做是否合适，如果需要改成类似 PaddleClas 那样每个预训练模型都暴露一个 API 的话我再改一下～

### CODE REVIEW

File: python/paddle/vision/models/mobilenetv3.py [hunk 1]
```diff
@@ -0,0 +1,516 @@
...
+        identity = x
+        x = self.avg_pool(x)
+        x = self.conv1(x)
+        x = self.relu(x)
+        x = self.conv2(x)
+        x = self.hardsigmoid(x)
+        return paddle.multiply(x=identity, y=x)
+
+
+class ConvBNLayer(nn.Layer):
```
Comment (LielinJiang) [inline python/paddle/vision/models/mobilenetv3.py]: 看看这些类能否复用已有的

File: python/paddle/vision/models/mobilenetv3.py [hunk 2]
```diff
@@ -0,0 +1,516 @@
...
+
+    def forward(self, x):
+        x = self.conv(x)
+        x = self.bn(x)
+        if self.act is not None:
+            x = self.act(x)
+        return x
+
+
+class InvertedResidual(nn.Layer):
```
Comment (LielinJiang) [inline python/paddle/vision/models/mobilenetv3.py:108]: 同上

Comment (SigureMo) [reply python/paddle/vision/models/mobilenetv3.py:108]: `InvertedResidual` 感觉不太容易复用，`MobileNetV2` 和 `MobileNetV3` 虽然都有这个模块，但是改动看起来还是蛮大的，复用起来可能麻烦些……

image

> MobileNetV3 paper, <https://arxiv.org/pdf/1905.02244>

File: python/paddle/vision/models/mobilenetv3.py (hunk 1 above)
Comment (SigureMo) [reply python/paddle/vision/models/mobilenetv3.py]: `ConvBNLayer` 看起来是能够复用的，而且随着各式各样 Vision Models 的添加，这种重复也会越来越大，就目前这些模型而言，类似的代码已经出现在了：

- inceptionv3.ConvBNLayer
- mobilenetv1.ConvBNLayer
- mobilenetv2.ConvBNLayer
- mobilenetv3.ConvBNLayer (当前模型)
- resnext.ConvBNLayer
- shufflenetv2.ConvBNLayer

我们是否可以参考 torchvision 里那样，在 `paddle.vision.ops` 里添加一个容易复用的通用模块，比如像 torchvision 的 `ConvNormActivation`

https://github.com/pytorch/vision/blob/e13206d9749e81fd8b3aec5e664f697a73febf9f/torchvision/ops/misc.py#L68-L124

这里 Norm 没有限制成 BN 貌似是为了兼容 LN 等 Norm 的模块，比如 [ConvNeXt](https://github.com/pytorch/vision/blob/e13206d9749e81fd8b3aec5e664f697a73febf9f/torchvision/models/convnext.py#L9)

不过如果要复用同一代码的话，由于现有各个模型内部的 `ConvBNLayer` 内部参数命名各不相同，所以可能涉及到更新现有权重的问题，整体做起来工程量可能不小。

如果上述方案是可以的话，我可以在本 PR 中实施创建可复用的模块 `ConvNormActivation`，并在 mobilenetv3 引用它。然后我会在下一个 PR 中在其余 5 个模型中引用它，并逐步更新参数。

@LielinJiang 请问这样可以嘛？

File: python/paddle/vision/models/mobilenetv3.py [hunk 3]
```diff
@@ -0,0 +1,516 @@
...
+
+
+def _make_divisible(v, divisor=8, min_value=None):
+    if min_value is None:
+        min_value = divisor
+    new_v = max(min_value, int(v + divisor / 2) // divisor * divisor)
+    # Make sure that round down does not go down by more than 10%.
+    if new_v < 0.9 * v:
+        new_v += divisor
+    return new_v
```
Comment (SigureMo) [inline python/paddle/vision/models/mobilenetv3.py]: 这样的常用的工具函数我是否可以像 torchvision 那样提取到一个单独的文件中？像 [torchvision.models._utils](https://github.com/pytorch/vision/blob/d367a01a18a3ae6bee13d8be3b63fd6a581ea46f/torchvision/models/_utils.py#L70) 那样，对应 paddle 的话可能类似 paddle.vision.models._utils 这样。目前它与 mobilenetv2 中的重复了。

File: python/paddle/vision/models/mobilenetv3.py (hunk 1 above)
Comment (LielinJiang) [reply python/paddle/vision/models/mobilenetv3.py]: 可以的~

File: python/paddle/vision/models/mobilenetv3.py [hunk 4]
```diff
@@ -0,0 +1,508 @@
...
+        return F.hardsigmoid(x, slope=self.slope, offset=self.offset)
+
+
+class SqueezeExcitation(nn.Layer):
+    def __init__(self, channels, reduction=4):
+        super().__init__()
+        self.avg_pool = nn.AdaptiveAvgPool2D(1)
+        self.conv1 = nn.Conv2D(
+            in_channels=channels,
+            out_channels=channels // reduction,
```
Comment (SigureMo) [inline python/paddle/vision/models/mobilenetv3.py]: @LielinJiang 

在对代码的修改和权重转换过程中，我发现 paddleclas 中目前的实现在 SE 模块中 squeeze 时通道数是直接除 4，而无论是 [keras.applications](https://github.com/keras-team/keras/blob/6bd6d0df856c2ebb386d37e953a2f2b2eb5abc3a/keras/applications/mobilenet_v3.py#L484-L489) 还是 [torchvision](https://github.com/pytorch/vision/blob/e13206d9749e81fd8b3aec5e664f697a73febf9f/torchvision/models/mobilenetv3.py#L110-L111) 都在除 4 后进一步使用 _make_divisible 保证其大小能够被 8 整除。比如说 scale=1 时的 mobilenetv3_large 在第一个 SE 模块中第一个卷积时理应输入 72 输出 _make_divisible(72//4)=24，但目前 paddleclas 实现是输出 72//4=18。唔我不太清楚这是不是因为疏漏导致的，但确实和 keras 和 pytorch 实现是不一样的。

这样的话，我们是不是应该放弃使用 paddleclas 里的预训练权重（因为如果修正模型结构的话，SE 那部分 shape 完全对不上），转而使用从 torchvision 现有模型转换的权重呢？不过 torchvision 目前是 small 和 large 都只有 scale=1.0 的权重……比 paddleclas 少不少……

Comment (LielinJiang) [reply python/paddle/vision/models/mobilenetv3.py]: 这部分最开始是和gluoncv对齐的。因为当时torch版本的实现无法训练出论文精度。我的建议是按照paddleclas的版本先合入，然后在文档中说明与torchvision版本的不同点

Comment (SigureMo) [reply python/paddle/vision/models/mobilenetv3.py]: 好哒～

Comment (LielinJiang) [reply python/paddle/vision/models/mobilenetv3.py]: 又和之前复现的同学探讨了下，确实clas的版本和其他框架版本都不一样（包括gluoncv）。所以感觉要不使用torchvision的版本吧，先把已有的权重转过来。缺少的权重我们后续再想办法

Comment (SigureMo) [reply python/paddle/vision/models/mobilenetv3.py]: 嗯，那我使用 torchvision 的试下～

Comment (LielinJiang) [reply python/paddle/vision/models/mobilenetv3.py]: 好的好的，麻烦了~

**Comment by SigureMo** [comment]
@LielinJiang

两个预训练权重已经从 torchvision 转换好啦（目前的链接是存在 aistudio 的链接）～（其他 scale 的也在本地做了前向传播对齐的验证，结构都是与 torchvision 一致的）

也将 `ConvBNLayer` 提取到了 `paddle.vision.ops.ConvNormActivation`，并在 mobilenetv3 中调用～

有时间可以麻烦 review 一下嘛～

File: python/paddle/vision/ops.py [hunk 5]
```diff
@@ -1297,3 +1298,54 @@ def forward(self, x, boxes, boxes_num, aligned=True):
...
+    """
+    Configurable block used for Convolution-Normalzation-Activation blocks.
+    This code is based on the pytorch code with modifications.
+    You can also see at https://github.com/pytorch/vision/blob/main/torchvision/ops/misc.py#L68
+    Args:
+        in_channels (int): Number of channels in the input image
+        out_channels (int): Number of channels produced by the Convolution-Normalzation-Activation block
+        kernel_size: (int, optional): Size of the convolving kernel. Default: 3
+        stride (int, optional): Stride of the convolution. Default: 1
+        padding (int, tuple or str, optional): Padding added to all four sides of the input. Default: None, in wich case it will calculated as ``padding = (kernel_size - 1) // 2 * dilation``
```
Comment (LielinJiang) [inline python/paddle/vision/ops.py]: 这里注释太长了，参考其他地方分下行吧

File: python/paddle/vision/ops.py [hunk 6]
```diff
@@ -1297,3 +1298,54 @@ def forward(self, x, boxes, boxes_num, aligned=True):
             output_size=self._output_size,
             spatial_scale=self._spatial_scale,
             aligned=aligned)
+
+
+class ConvNormActivation(Sequential):
+    """
+    Configurable block used for Convolution-Normalzation-Activation blocks.
+    This code is based on the pytorch code with modifications.
```
Comment (LielinJiang) [inline python/paddle/vision/ops.py]: pytorch->torchvision

File: python/paddle/vision/ops.py (hunk 5 above)
Comment (SigureMo) [reply python/paddle/vision/ops.py]: 已修改～

File: python/paddle/vision/ops.py (hunk 6 above)
Comment (SigureMo) [reply python/paddle/vision/ops.py]: 已修改～

**Review by LielinJiang — DISMISSED** [review]

**Review by TCChenlong — DISMISSED** [review]
LGTM

File: python/paddle/vision/models/mobilenetv3.py [hunk 7]
```diff
@@ -0,0 +1,445 @@
...
+__all__ = []
+
+model_urls = {
+    "mobilenet_v3_small_x1.0":
+    ("https://bj.bcebos.com/v1/ai-studio-online/1552a31af0504d5da16291204f2964309ca1f02ff66c485488da507c6117576d",
+     "34fe0e7c1f8b00b2b056ad6788d0590c"),
+    "mobilenet_v3_large_x1.0":
+    ("https://bj.bcebos.com/v1/ai-studio-online/7a74ecf7e5d54811a11e10789568c1845ec1b77b9e34424c82378babe6ffbdc0",
+     "118db5792b4e183b925d8e8e334db3df"),
+}
```
Comment (SigureMo) [inline python/paddle/vision/models/mobilenetv3.py]: @LielinJiang 可以麻烦有时间上传下这两个权重嘛～

Comment (LielinJiang) [reply python/paddle/vision/models/mobilenetv3.py]: https://paddle-hapi.bj.bcebos.com/models/mobilenet_v3_large_x1.0.pdparams
https://paddle-hapi.bj.bcebos.com/models/mobilenet_v3_small_x1.0.pdparams

**Comment by SigureMo** [comment]
@TCChenlong @LielinJiang 权重链接修改了下并过了下 CI，可以麻烦重新 review 下嘛？

**Review by LielinJiang — DISMISSED** [review]

**Review by TCChenlong — DISMISSED** [review]
LGTM

**Comment by SigureMo** [comment]
@XiaoguangHu01 有时间可以 review 一下嘛？

File: python/paddle/vision/ops.py [hunk 8]
```diff
@@ -36,6 +36,7 @@
     'PSRoIPool',
     'roi_align',
     'RoIAlign',
+    'ConvNormActivation',
```
Comment (XiaoguangHu01) [inline python/paddle/vision/ops.py]: 这个接口作为内部实现吧，先不用作为公开接口对外暴露

File: python/paddle/tests/test_pretrained_model.py [hunk 9]
```diff
@@ -61,6 +61,7 @@ def test_models(self):
         arches = [
             'mobilenet_v1',
             'mobilenet_v2',
+            'mobilenet_v3_small',
```
Comment (XiaoguangHu01) [inline python/paddle/tests/test_pretrained_model.py:64]: 这里缺少mobilenet_v3_large，是因为mobilenet_v3_large模型太大吗？

File: python/paddle/vision/models/_utils.py [hunk 10]
```diff
@@ -0,0 +1,32 @@
+# Copyright (c) 2022 PaddlePaddle Authors. All Rights Reserved.
```
Comment (XiaoguangHu01) [inline python/paddle/vision/models/_utils.py]: 文件名不用加下划线前缀_utils.py -> utils.py

File: python/paddle/tests/test_pretrained_model.py (hunk 9 above)
Comment (SigureMo) [reply python/paddle/tests/test_pretrained_model.py:64]: 嗯嗯，之前是看到 [#37307](https://github.com/PaddlePaddle/Paddle/pull/37307) 暂时移除了较大的模型，不过看起来 mobilenet_v3_large (21MB) 相对来说也不算大，刚刚已经加上了～

File: python/paddle/vision/models/_utils.py (hunk 10 above)
Comment (SigureMo) [reply python/paddle/vision/models/_utils.py]: 已修改～

File: python/paddle/vision/ops.py (hunk 8 above)
Comment (SigureMo) [reply python/paddle/vision/ops.py]: 已经取消暴露～

**Review by TCChenlong — APPROVED** [review]
LGTM

**Review by LielinJiang — APPROVED** [review]

**Review by XiaoguangHu01 — APPROVED** [review]
LGTM

File: python/paddle/vision/ops.py [hunk 11]
```diff
@@ -1297,3 +1297,57 @@ def forward(self, x, boxes, boxes_num, aligned=True):
             output_size=self._output_size,
             spatial_scale=self._spatial_scale,
             aligned=aligned)
+
+
+class ConvNormActivation(Sequential):
+    """
+    Configurable block used for Convolution-Normalzation-Activation blocks.
+    This code is based on the torchvision code with modifications.
+    You can also see at https://github.com/pytorch/vision/blob/main/torchvision/ops/misc.py#L68
```
Comment (qingqing01) [inline python/paddle/vision/ops.py:1306]: 需要修改文档，不能重复，并且去掉参考链接~

Comment (SigureMo) [reply python/paddle/vision/ops.py:1306]: 「不能重复」是指？另外这个 API 不是一个会对外暴露的 API，目前不会在官网生成文档～
